### DISCUSSION TITLE
Callback before submit

### DISCUSSION THREAD

**Discussion by babur001** [react-hook-form/react-hook-form#6844] [Ideas]
I always have a problem when i use mask, ex: `+998 (90) 777 77 77` number stays as it is.

It would be a great to have some callback to change the input whenever submitted

Example when used Mask:

```jsx
<InputMask
  mask="+\9\98 99 999 99 99"
  {...register("phone", {}, (data) => {
    return data.split(" ").join("");
  })}
>
  {(inputProps) => (
    <input
      type="text"
      placeholder="+998 (__)  ___-__-__"
      className="input | m-0 | w-full | input-bordered"
      {...inputProps}
    />
  )}
</InputMask>
```
And it will get the clear output of `+998907277777`, which will be a valid data to send to an API.

Or is there an alternative solution already for that?

For now, I am doing this logic directly in `handleSubmit(login)` function:

```jsx
async function login(data) {
  const cleanPhone = data.phone.split(" ").join("");

  const formattedData = {
    ...data,
    phone: cleanPhone,
  };

  // Send the formattedData
}
```

**Comment by JamesWong1999** [comment]
Use `Controller` instead?  It's a controlled component, which kind of explains why the last keystroke is missing.

here is a group of examples: https://codesandbox.io/s/react-hook-form-v7-controller-5h1q5

**Reply by xkomiks** [reply]
@JamesWong1999 Still, I would like to see the callback before submit function.

An example format of values could be like this

```js
const values = {
  [fieldType]: { 
    [fieldName]: value,
    ...
   },
 ...
}
```

**Reply by Moshyfawn** [reply]
You can simply do
```js
handleSubmit(formatData(onSubmit))
```

**Reply by xkomiks** [reply]
I use such a workaround only for data normalization (`trimStrings`, `removeEmptyStrings`), but because it's only values without type, I can't format them.

**Reply by Moshyfawn** [reply]
The showed example is not a workaround but rather the way to format your form data. I'm afraid I don't understand what you mean by
> because it's only values without type, I can't format them.

Can you elaborate on that?

Another way to format the input's value is to use the setValueAs `register` option callback like
```js
setValueAs: value => formatValue(width)
```
But it **happens on value change rather on the submit event**

To double down on the @JamesWong1999's valid point, with the controlled by design inputs like `InputMask`, consider using `Controller` / `useController` API

**Reply by babur001** [reply]
> The showed example is not a workaround but rather the way to format your form data. I'm afraid I don't understand what you mean by
> 
> > because it's only values without type, I can't format them.
> 
> Can you elaborate on that?
> 
> Another way to format the input's value is to use the setValueAs `register` option callback like
> 
> ```js
> setValueAs: value => formatValue(width)
> ```
> 
> But it **happens on value change rather on the submit event**
> 
> To double down on the @JamesWong1999's valid point, with the controlled by design inputs like `InputMask`, consider using `Controller` / `useController` API

Why i don not want it to be serialized inside onSubmit is when i use non-serialized input outside of onSubmit is not a valid one so need to wrap it onto `serialize(input)` function, cannot directly properly access input with `watch('input-name')`I have to do some extra work with wrapping with func. 

This is what I wanted to achieve, accessing valid input wherever I want:

```jsx
setValueAs: value => formatValue(width)

... 

watch('name') // '+998907277777' -> valid input!

const onSubmit = (form: TForm) => {
 consonle.log(form.name)  // '+998907277777' valid input!
}
```
@Moshyfawn Thank you, set it as an answer 👍
